File: backend/src/main/java/io/github/artemy/osipov/shop/service/report/ReportPresent.java
package io.github.artemy.osipov.shop.service.report;

import io.github.artemy.osipov.shop.service.candy.Candy;
import io.github.artemy.osipov.shop.service.present.Present;
import lombok.Value;

import java.math.BigDecimal;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Collectors;

@Value
public class ReportPresent {

    String name;
    BigDecimal price;
    List<Item> items;

    public ReportPresent(Present present, Collection<Candy> candies) {
        this.name = present.getName();
        this.price = present.getPrice();
        this.items = buildItems(present, candies);
    }

    private static List<Item> buildItems(Present present, Collection<Candy> candies) {
        Map<UUID, Candy> idToCandy = candies.stream()
                .collect(Collectors.toMap(
                        Candy::getId,
                        Function.identity()
                ));
        return present.getItems()
                .stream()
                .map(i -> new Item(
                        Objects.requireNonNull(
                                idToCandy.get(i.getCandyId())
                        ),
                        i.getCount()
                ))
                .sorted(Comparator.comparing((Item i) -> i.getCandy().getOrder()))
                .collect(Collectors.toList());
    }

    public BigDecimal computeCost() {
        return items.stream()
                .map(item -> item.getCandy().getPrice().multiply(BigDecimal.valueOf(item.getCount())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    @Value
    public static class Item {
        Candy candy;
        int count;
    }
}

File: backend/src/main/kotlin/io/github/artemy/osipov/shop/service/report/ReportPresent.kt
package io.github.artemy.osipov.shop.service.report

import io.github.artemy.osipov.shop.service.candy.Candy
import io.github.artemy.osipov.shop.service.present.Present

import java.math.BigDecimal
import java.util.Comparator
import java.util.UUID
import java.util.function.Function
import java.util.stream.Collectors

data class ReportPresent(
    val name: String,
    val price: BigDecimal,
    val items: List<Item>
) {
    constructor (present: Present, candies: Collection<Candy>) : this(
        name = present.name,
        price = present.price,
        items = buildItems(present, candies)
    )

    private companion object {
        fun buildItems(present: Present, candies: Collection<Candy>): List<Item> {
            val idToCandy: Map<UUID, Candy> = candies.stream()
                .collect(
                    Collectors.toMap(
                        Candy::id,
                        Function.identity()
                    )
                )
            return present.items
                .stream()
                .map { i -> Item(idToCandy[i.candyId]!!, i.count) }
                .sorted(Comparator.comparing { i -> i.candy.order })
                .collect(Collectors.toList())
        }
    }

    fun computeCost(): BigDecimal {
        return items.stream()
            .map { item -> item.candy.price.multiply(BigDecimal(item.count)) }
            .reduce(BigDecimal.ZERO, BigDecimal::add)
    }

    data class Item(
        val candy: Candy,
        val count: Int
    )
}
